MvvmCross/Core/Core/Navigation/MvxNavigationService.cs

```diff
@@ -171,78 +173,137 @@ private static bool TryGetRoute(string url, out KeyValuePair<Regex, Type> entry)
             return TryGetRoute(path, out entry);
         }
 
-        public async Task Navigate<TViewModel>() where TViewModel : IMvxViewModel
+        public Task Navigate<TViewModel>() where TViewModel : IMvxViewModel
         {
             var args = new NavigateEventArgs(typeof(TViewModel));
             OnBeforeNavigate(this, args);
             _viewDispatcher.ShowViewModel(new MvxViewModelRequest<TViewModel>());
             OnAfterNavigate(this, args);
+
+            return Task.FromResult(true);
         }
 
-        public async Task<bool> Close(IMvxViewModel viewModel)
+        public Task<bool> Close(IMvxViewModel viewModel)
         {
             var args = new NavigateEventArgs();
             OnBeforeClose(this, args);
             var close = _viewDispatcher.ChangePresentation(new MvxClosePresentationHint(viewModel));
             OnAfterClose(this, args);
-            return close;
+
+            return Task.FromResult(close);
         }
 
-        public Task<TResult> Navigate<TViewModel, TParameter, TResult>(TParameter param)
+        public async Task<TResult> Navigate<TViewModel, TParameter, TResult>(TParameter param)
             where TViewModel : IMvxViewModel<TParameter, TResult>
             where TParameter : class
             where TResult : class
         {
-            throw new NotImplementedException();
+            var args = new NavigateEventArgs(typeof(TViewModel));
+            OnBeforeNavigate(this, args);
+
+			var viewModel = (IMvxViewModel<TParameter, TResult>)Mvx.IocConstruct<TViewModel>();
+            var request = new MvxViewModelInstanceRequest(viewModel);
+
+            var tcs = new TaskCompletionSource<TResult>();
+            viewModel.SetClose(tcs);
+
+            _viewDispatcher.ShowViewModel(request);
+            await viewModel.Initialize(param);
+
+            OnAfterNavigate(this, args);
+
+            try
+            {
+                return await tcs.Task;
+            }
+            catch (OperationCanceledException ex)
+            {
+                return default(TResult);
```

Comment:
No I think that's fine as long as `null` is only returned in the case of cancellation.